detox/e2e/test/channels/join_from_permalink.e2e.js

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// *******************************************************************
+// - [#] indicates a test step (e.g. # Go to a screen)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element testID when selecting an element. Create one if none.
+// *******************************************************************
+
+import {ChannelSidebar} from '@support/ui/component';
+import {ChannelScreen} from '@support/ui/screen';
+
+import {Setup, Team, User} from '@support/server_api';
+import {serverUrl} from '@support/test_config';
+
+describe('Channels', () => {
+    let testChannel;
+    let testTeam;
+
+    beforeAll(async () => {
+        const {channel: channel1, team, user} = await Setup.apiInit();
+        testChannel = channel1;
+        testTeam = team;
+
+        await ChannelScreen.open(user);
+    });
+
+    it('should join channel from a permalink', async () => {
+        // # Go to the Town Square channel
+        const {channelDrawerButton, channelNavBarTitle} = ChannelScreen;
+        await channelDrawerButton.tap();
+        await ChannelSidebar.toBeVisible();
+        let channelItem = ChannelSidebar.getChannelByDisplayName('Town Square');
+        await channelItem.tap();
+        await expect(channelNavBarTitle).toHaveText('Town Square');
+
+        // # There's no way to get a channel permalink on mobile so we make one
+        // manually
+        const permalink = `${serverUrl}/${testTeam.name}/channels/${testChannel.name}`;
+
+        // # Post the permalink to the test channel in Town Square
+        const {postInput, sendButton} = ChannelScreen;
+        await expect(postInput).toBeVisible();
+        await expect(sendButton).not.toExist();
+        await postInput.tap();
+        await postInput.typeText(permalink);
+        await expect(sendButton).toBeVisible();
+        await sendButton.tap();
+        await expect(element(by.text(permalink))).toBeVisible();
+
+        // # Create another user in the same team, log in and go to town square
+        const {user: otherUser} = await User.apiCreateUser({prefix: 'TestUser'});
+        await Team.apiAddUserToTeam(otherUser.id, testTeam.id);
+        await ChannelScreen.logout();
+        await ChannelScreen.open(otherUser);
+        await channelDrawerButton.tap();
+        await ChannelSidebar.toBeVisible();
+        channelItem = ChannelSidebar.getChannelByDisplayName('Town Square');
+        await channelItem.tap();
+        await expect(channelNavBarTitle).toHaveText('Town Square');
+
+        // # As this new user, Tap the permalink we posted earlier
+        const permalinkPost = element(by.text(permalink));
+        await permalinkPost.tap({x: 5, y: 10});
```

Comment:
@saturninoabril selecting by text gives me the whole post, and tapping the post gives me the post view. Here I've worked around it by guessing the link's co-ordinates withing the post. This is discouraged by detox because depending on device dimensions or orientation, we could end up tapping the wrong place.

Reply:
Honestly, I'm good with this approach but I'd limit to `{x: 5, y: 5}` so it's still within the element boundary (if rendered either one or two-liner text). One option is to add specific testID to URL link (as mentioned by Joseph) but what I like with this approach (specifically the `by.text`) is it's checking the actual text visible to the user.

Reply:
My take is we should separate concerns about querying the element and verifying the element's attribute. Text values can change and we can always add an explicit verification in the test (which is my preference) like `.toHaveText()` or `.toHaveLabel()`

Reply:
I like the idea of querying an element similarly on how the users would find them. So, if no other attribute other than the text itself is going to verify for the element, then I'll go with `by.text` or `by.label` whenever possible since it's straightforward.

Reply:
I understand but isn't that a bit contradictory to the guides where we encourage testID as priority over other element queries? https://developers.mattermost.com/contribute/mobile/e2e/guide-for-writing/ If we're going by text and label, then we should probably update the docs to reflect that priority is text/label then testID

Reply:
I've been looking into adding a testID to the markdown link. Detox is unable to tap the markdown link because it thinks the element is not visible. From what I can make out, the detox may think an element is not visible if some other element is rendered on top of it (even if the element on top is transparent). One way to confirm this is to check the the 3D hierarchy through xcode. I'm still not setup on OSX so I can't confirm this. 
I've tried injecting the testID into the child components, but detox still thinks they're not visible. So I'm not sure what's going on.

Reply:
@ashishbhate I see upon further investigation, markdown_link is wrapped in Text container. I believe Text container does not have testID prop. Detox can only query testID from components that already have it, e.g. View, TouchableOpacity, etc. or descendants. Either wrap it with a _testID'able_ container if possible, or go with the original approach of coordinates, since we know it's not possible with Text container. Thanks for looking into.

Reply:
Chances are that wrapping the Text in a View container that has testID you'll break the layout and cause other potential issues. Stick with the coordinates or use another type of selector instead of testID